### PR TITLE
[Auto Techsupport] Event driven Techsupport Changes

### PR DESCRIPTION
**Pull request by vivekreddynv**
#### Why I did it

Changes required for feature "Event Driven TechSupport Invocation & CoreDump Mgmt" HLD.

#### How to verify it

```
##### No Core files and ts dumps currently
admin@sonic:~$ ls /var/core/
admin@sonic:~$ ls /var/dump
ls: cannot access '/var/dump': No such file or directory

#### verify auto-techsupport status
admin@sonic:~$ show auto-techsupport global
admin@sonic:~$ show auto-techsupport global
STATE      RATE LIMIT INTERVAL    MAX TECHSUPPORT SIZE    MAX CORE SIZE  SINCE
-------  ---------------------  ----------------------  ---------------  ----------
enabled                    180                      10                5  2 days ago


#### Kill a critical Process t and trigger a coredump
admin@sonic:~$ docker exec -it snmp ps -aux
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root          23  1.6  0.5 138316 42296 pts/0    Sl   19:09   0:15 python3 -m sonic_ax_impl
...............

admin@sonic:~$ docker exec -it snmp kill -11 23

#### Coredump is created
admin@sonic:~$ ls /var/core/
python3.1629401152.23.core.gz 

#### Techsupport Dump creation in progress
admin@sonic:~$ ls /var/dump/
sonic_dump_sonic_20210819_192558  sonic_dump_sonic_20210819_192558.tar

admin@sonic:~$ ps -aux | grep coredump_gen
root       17823  0.1  0.2  30960 16736 ?        S    19:25   0:00 python3 /usr/local/bin/coredump_gen_handler.py python3.1629401152.23.core.gz

#### Wait until the techsupport dump execution has finished
admin@sonic:~$ ls /var/dump/
sonic_dump_sonic_20210819_192558.tar.gz


admin@r-lionfish-16:~$ show auto-techsupport history
Techsupport Dump                                 Triggered By                   Critical Process
-----------------------------------------------  -----------------------------  ------------------
sonic_dump_sonic_20210819_192558.tar.gz  python3.1629401152.23.core.gz  snmp-subagent
```

Note on changes made to supervisor-proc-exit-listener script:
Changes were made are backward compatible with python2. Tested this on docker running python3 & python2
```
admin@sonic:~$ docker exec -it snmp ps -aux
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root           1  0.5  0.3  35676 24316 pts/0    Ss+  21:43   0:00 /usr/bin/pyth
root           9  0.1  0.3  43608 24888 pts/0    S    21:43   0:00 python3 /usr/
root          17  0.0  0.0 225856  5616 pts/0    Sl   21:43   0:00 /usr/sbin/rsy
Debian-+      21  0.9  0.1  32932 12524 pts/0    S    21:43   0:00 /usr/sbin/snm
root          23  8.4  0.4 133116 36764 pts/0    Sl   21:43   0:06 python3 -m so
root          26  0.0  0.0  11248  3036 pts/1    Rs+  21:44   0:00 ps -aux
admin@r-lionfish-16:~$ docker exec -it restapi ps -aux
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root           1  0.5  0.2  59112 21552 pts/0    Ss+  21:43   0:00 /usr/bin/pyth
root           9  0.1  0.2  84932 22932 pts/0    S    21:43   0:00 python2 /usr/
root          14  0.0  0.0 262988  3596 pts/0    Sl   21:43   0:00 /usr/sbin/rsy
root          20  0.0  0.0  17964  2948 pts/0    S    21:43   0:00 bash /usr/bin
root          31  0.0  0.0   4188   660 pts/0    S    21:44   0:00 sleep 60
root          32  0.0  0.0  36636  2840 pts/1    Rs+  21:44   0:00 ps -aux
admin@sonic:~$ docker exec -it restapi kill -11 20
admin@sonic:~$ docker exec -it snmp kill -11 23
admin@sonic:~$ redis-cli -n 6 hgetall "AUTO_TECHSUPPORT|FEATURE_PROC_INFO"
1) "restapi;restapi"
2) "20;bash"
3) "snmp;snmp-subagent"
4) "23;python3"
```

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106

### CODE REVIEW

File: files/scripts/supervisor-proc-exit-listener [hunk 1]
```diff
@@ -124,20 +167,26 @@ def main(argv):
     # Transition from ACKNOWLEDGED to READY
     childutils.listener.ready()
 
+    # STATE_DB dbconnector
+    dbconn = DBConnector('STATE_DB', 0)
```
[the remark this answers is not in the document]
Comment (nazariig) [reply files/scripts/supervisor-proc-exit-listener:171]: @vivekreddynv meant a python constant for 'STATE_DB'

[the remark this answers is not in the document]
Comment (vivekreddynv) [reply files/scripts/supervisor-proc-exit-listener:171]: Oh, if inside the script, NO. 

I can see if there is a named def in the imported modules or in swsscommon. Is that what you've suggested

File: files/scripts/supervisor-proc-exit-listener [hunk 2]
```diff
@@ -31,6 +32,48 @@ SELECT_TIMEOUT_SECS = 1.0
...
+            comm_name = PID_COMM_FILE_MAP.get(pid, "")
+            if not comm_name:
+                comm_name = "<unknown>"
+            # Create an entry in the Table
+            field = "{};{}".format(container_name, supervisor_proc_name)
+            value = "{};{}".format(pid, comm_name)
+            dbconn.hset("AUTO_TECHSUPPORT|FEATURE_PROC_INFO", field, value)
+            del PID_COMM_FILE_MAP[pid]
+        elif event == "PROCESS_STATE_RUNNING":
+            rc, stdout, stderr = subprocess_exec(["cat", "/proc/{}/comm".format(pid)])
```
[the remark this answers is not in the document]
Comment (vivekreddynv) [reply files/scripts/supervisor-proc-exit-listener:68]: `check_output() will fail with a CalledProcessError if the requested operation produces a non-zero return code`. 

With subprocess.run(), i can capture the return code and act upon it.

File: files/scripts/supervisor-proc-exit-listener [hunk 3]
```diff
@@ -31,6 +32,48 @@ SELECT_TIMEOUT_SECS = 1.0
...
+            if not rc:
+                PID_COMM_FILE_MAP[pid] = stdout.strip()  # Update the comm<->pid map
+            # Delete the Event Entry, since the process has restarted
+            field = "{};{}".format(container_name, supervisor_proc_name)
+            dbconn.hdel("AUTO_TECHSUPPORT|FEATURE_PROC_INFO", field)
+    except Exception as e:
+        syslog.syslog(syslog.LOG_NOTICE, "handle_proc_event method failed, ERR: {}".format(e))
```
[the remark this answers is not in the document]
Comment (vivekreddynv) [reply files/scripts/supervisor-proc-exit-listener:75]: Handled

File: files/scripts/supervisor-proc-exit-listener [hunk 4]
```diff
@@ -31,6 +32,48 @@ SELECT_TIMEOUT_SECS = 1.0
 # Alerting message will be written into syslog in the following interval
 ALERTING_INTERVAL_SECS = 60
 
+# Saves the mapping between PID and the command/executable name
+PID_COMM_FILE_MAP = dict()
```
[the remark this answers is not in the document]
Comment (vivekreddynv) [reply files/scripts/supervisor-proc-exit-listener:36]: No updating this for backward compatibility reasons as specified in the comment above

File: src/sonic-yang-models/tests/yang_model_tests/tests_config/auto_techsupport.json [hunk 5]
```diff
@@ -0,0 +1,85 @@
+{
+    "AUTO_TECHSUPPORT_TEST": {
```
Comment (nazariig) [inline src/sonic-yang-models/tests/yang_model_tests/tests_config/auto_techsupport.json:2]: @vivekreddynv `AUTO_TECHSUPPORT_FEATURE` data is missing.

**Review by nazariig — APPROVED** [review]

**Pull request closed without merging**